esp-hal/src/dma/buffers.rs

```diff
@@ -4,25 +4,252 @@ use core::{
+            4
+        } else {
+            // OUT transfers have no alignment requirements, except for ESP32, which is
+            // described below.
+            if cfg!(esp32) {
+                // SPI DMA: Burst transmission is supported. The data size for
+                // a single transfer must be four bytes aligned.
+                // I2S DMA: Burst transfer is supported. However, unlike the
+                // SPI DMA channels, the data size for a single transfer is
+                // one word, or four bytes.
+                4
+            } else {
+                1
+            }
+        }
+    }
+}
+
+const fn max(a: usize, b: usize) -> usize {
+    if a > b {
+        a
+    } else {
+        b
+    }
+}
+
+impl BurstConfig {
+    delegate::delegate! {
+        #[cfg(psram_dma)]
+        to self.internal {
+            pub(super) const fn min_dram_alignment(self, direction: TransferDirection) -> usize;
+            pub(super) fn is_burst_enabled(self) -> bool;
+        }
+    }
+
+    /// Calculates an alignment that is compatible with the current burst
+    /// configuration.
+    ///
+    /// This is an over-estimation so that Descriptors can be safely used with
+    /// any DMA channel in any direction.
+    pub const fn min_compatible_alignment(self) -> usize {
+        let in_alignment = self.min_dram_alignment(TransferDirection::In);
+        let out_alignment = self.min_dram_alignment(TransferDirection::Out);
+        let alignment = max(in_alignment, out_alignment);
+
+        #[cfg(psram_dma)]
+        let alignment = max(alignment, self.external as usize);
+
+        alignment
+    }
+
+    /// Calculates a chunk size that is compatible with the current burst
+    /// configuration's alignment requirements.
+    ///
+    /// This is an over-estimation so that Descriptors can be safely used with
+    /// any DMA channel in any direction.
+    pub const fn max_compatible_chunk_size(self) -> usize {
+        4096 - self.min_compatible_alignment()
+    }
+
+    fn min_alignment(self, _buffer: &[u8], direction: TransferDirection) -> usize {
+        let alignment = self.min_dram_alignment(direction);
+
+        cfg_if::cfg_if! {
+            if #[cfg(psram_dma)] {
+                let mut alignment = alignment;
+                if is_valid_psram_address(_buffer.as_ptr() as usize) {
+                    alignment = max(alignment, self.external.min_psram_alignment(direction));
+                }
+            }
+        }
+
+        alignment
+    }
+
+    // Note: this function ignores address alignment as we assume the buffers are
+    // aligned.
+    fn max_chunk_size_for(self, buffer: &[u8], direction: TransferDirection) -> usize {
+        4096 - self.min_alignment(buffer, direction)
+    }
+
+    fn is_buffer_aligned(self, buffer: &[u8], direction: TransferDirection) -> bool {
+        let alignment = self.min_alignment(buffer, direction);
+        buffer.as_ptr() as usize % alignment == 0
+    }
```

Comment:
You don't check the length as well?

Reply:
I'm not sure why I didn't, I think I managed to conclude that we don't need to. In one direction it doesn't matter, in the other the data is padded. But I agree that this looks fishy and it needs to be revisited, and at the very least explained.

Reply:
I remember now, my conclusion was that we don't need to check length. We need the capacity to be a multiple of the burst size, but the length has no such requirements that I could see.